server/odc-service/src/main/java/com/oceanbase/odc/service/sqlcheck/rule/MySQLAffectedRows.java

```diff
@@ -0,0 +1,166 @@
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import org.springframework.jdbc.core.JdbcOperations;
+
+import com.oceanbase.odc.core.shared.constant.DialectType;
+import com.oceanbase.odc.service.sqlcheck.SqlCheckContext;
+import com.oceanbase.odc.service.sqlcheck.SqlCheckRule;
+import com.oceanbase.odc.service.sqlcheck.SqlCheckUtil;
+import com.oceanbase.odc.service.sqlcheck.model.CheckViolation;
+import com.oceanbase.odc.service.sqlcheck.model.SqlCheckRuleType;
+import com.oceanbase.tools.sqlparser.statement.Statement;
+import com.oceanbase.tools.sqlparser.statement.delete.Delete;
+import com.oceanbase.tools.sqlparser.statement.insert.Insert;
+import com.oceanbase.tools.sqlparser.statement.update.Update;
+
+import lombok.NonNull;
+
+/**
+ * {@link MySQLAffectedRows}
+ *
+ * @author yiminpeng
+ * @version 1.0
+ * @date 2024-08-01 18:18
+ */
+public class MySQLAffectedRows implements SqlCheckRule {
+
+    private final Integer maxSQLAffectedRows;
+
+    private final JdbcOperations jdbcOperations;
+
+    public MySQLAffectedRows(@NonNull Integer maxSQLAffectedRows, JdbcOperations jdbcOperations) {
+        this.maxSQLAffectedRows = maxSQLAffectedRows <= 0 ? 0 : maxSQLAffectedRows;
+        this.jdbcOperations = jdbcOperations;
+    }
+
+    /**
+     * Get the rule type
+     */
+    @Override
+    public SqlCheckRuleType getType() {
+        return SqlCheckRuleType.RESTRICT_SQL_AFFECTED_ROWS;
+    }
+
+    /**
+     * Execution rule check
+     */
+    @Override
+    public List<CheckViolation> check(@NonNull Statement statement, @NonNull SqlCheckContext context) {
+
+        if (statement instanceof Insert || statement instanceof Update || statement instanceof Delete) {
+            if (maxSQLAffectedRows == 0) {
+                return Collections.emptyList();
+            }
+            String originSql = statement.getText();
+            long affectedRows = executeExplain(originSql, jdbcOperations);
+            if (affectedRows > maxSQLAffectedRows) {
+                return Collections.singletonList(SqlCheckUtil
+                        .buildViolation(statement.getText(), statement, getType(), new Object[] {}));
+            } else {
+                return Collections.emptyList();
+            }
+        }
+        return Collections.emptyList();
+    }
+
+    /**
+     * Get supported database types
+     */
+    @Override
+    public List<DialectType> getSupportsDialectTypes() {
+        return Arrays.asList(DialectType.MYSQL, DialectType.OB_MYSQL);
+    }
+
+    /**
+     * execute 'explain' statement
+     *
+     * @param originSql target sql
+     * @param jdbc jdbc Object
+     * @return affected rows
+     */
+    private long executeExplain(String originSql, JdbcOperations jdbc) {
+
+        if (originSql == null || jdbc == null) {
+            return 0;
+        }
+        String explainSql = "EXPLAIN " + originSql;
+
+        try {
```

Comment:
It's kind of hard to understand the logic. Could you explain it in more detail? Including but not limited to
- Why we need the `ifFindValue` flag?
- what does `rs.getString(1)` actually represent?
- Why when only `affectedRows != 0 ` then you stop?
- Why `rowNum  > 2`?
- so on...

Reply:
These are all related to the result set executed by explain on odc. These operations are all to get the value of "est.rows" in the result.

Reply:
Maybe I can achieve it in another way.

Reply:
better if encapsulate 'extract explain result' into Utils, e.g OBExplainUtils . may reuse some method in diagnose module, i'm not sure.

Reply:
Maybe we can extract it together after we improve the support for Oracle database next time.